src/umpire/strategy/ResourceAwarePool.cpp

```diff
@@ -0,0 +1,549 @@
+        }
+#endif
+        ret = aligned_allocate(size); // Will Poison
+      } catch (...) {
+        UMPIRE_LOG(Error,
+                   "Caught error allocating new chunk, giving up free chunks and "
+                   "retrying...");
+        release();
+        try {
+          ret = aligned_allocate(size); // Will Poison
+          UMPIRE_LOG(Debug, "memory reclaimed, chunk successfully allocated.");
+        } catch (...) {
+          UMPIRE_LOG(Error, "recovery failed.");
+          throw;
+        }
+      }
+
+      m_actual_bytes += size;
+      m_releasable_bytes += size;
+      m_releasable_blocks++;
+      m_total_blocks++;
+      m_actual_highwatermark = (m_actual_bytes > m_actual_highwatermark) ? m_actual_bytes : m_actual_highwatermark;
+
+      void* chunk_storage{m_chunk_pool.allocate()};
+      chunk = new (chunk_storage) Chunk{ret, size, size, r};
+    } else {
+      chunk = (*best).second;
+      chunk->m_resource = r;
+      m_free_map.erase(best);
+    }
+  }
+
+  UMPIRE_LOG(Debug, "Using chunk " << chunk << " with data " << chunk->data << " and size " << chunk->size
+                                   << " for allocation of size " << rounded_bytes);
+
+  if ((chunk->size == chunk->chunk_size) && chunk->free) {
+    m_releasable_bytes -= chunk->chunk_size;
+    m_releasable_blocks--;
+  }
+
+  void* ret = chunk->data;
+  m_used_map.insert(std::make_pair(ret, chunk));
+
+  chunk->free = false;
+
+  if (rounded_bytes != chunk->size) {
+    std::size_t remaining{chunk->size - rounded_bytes};
+    UMPIRE_LOG(Debug, "Splitting chunk " << chunk->size << "into " << rounded_bytes << " and " << remaining);
+
+    void* chunk_storage{m_chunk_pool.allocate()};
+    Chunk* split_chunk{new (chunk_storage)
+                           Chunk{static_cast<char*>(ret) + rounded_bytes, remaining, chunk->chunk_size, r}};
+    auto old_next = chunk->next;
+    chunk->next = split_chunk;
+    split_chunk->prev = chunk;
+    split_chunk->next = old_next;
+
+    if (split_chunk->next)
+      split_chunk->next->prev = split_chunk;
+
+    chunk->size = rounded_bytes;
+    split_chunk->size_map_it = m_free_map.insert(std::make_pair(remaining, split_chunk));
+    split_chunk->free = true;
+  }
+
+  m_current_bytes += rounded_bytes;
+
+  UMPIRE_UNPOISON_MEMORY_REGION(m_allocator, ret, bytes);
+  return ret;
+}
+
+void ResourceAwarePool::deallocate(void* ptr, std::size_t size)
+{
+  auto r = getResource(ptr);
+
+  UMPIRE_LOG(Warning, fmt::format("The ResourceAwarePool requires a Camp resource. Calling deallocate with: {}.",
+                                  camp::resources::to_string(r)));
+
+  deallocate_resource(r, ptr, size);
+}
+
+void ResourceAwarePool::do_deallocate(Chunk* chunk, void* ptr) noexcept
+{
+  UMPIRE_POISON_MEMORY_REGION(m_allocator, ptr, chunk->size);
+  UMPIRE_USE_VAR(ptr);
+  chunk->free = true;
+
+  // Removing chunk from pending
+  for (auto it = m_pending_map.begin(); it != m_pending_map.end();) {
```

Comment:
Why not call m_pending_map.find()? Or even better, in the allocate_resource function you already have the iterator and could pass that in to this function.